src/igraph_eulerian.c

```diff
@@ -0,0 +1,660 @@
+
+    }
+
+    if (cluster_count > 1) {
+        *has_path = 0;
+        *has_cycle = 0;
+        igraph_vector_destroy(&csize);
+        igraph_vector_destroy(&check_for_self_loops);
+        igraph_vector_destroy(&cluster_member);
+        IGRAPH_FINALLY_CLEAN(3);
+
+        return IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    vsize = igraph_vector_size(&check_for_self_loops);
+    for (i = 0; i < vsize; i++) {
+        igraph_vector_int_t *incedges = igraph_inclist_get(&il, VECTOR(check_for_self_loops)[i]);
+        long int nc = igraph_vector_int_size(incedges);
+        if (nc > 0) {
+            *has_path = 0;
+            *has_cycle = 0;
+            igraph_vector_destroy(&csize);
+            igraph_inclist_destroy(&il);
+            igraph_vector_destroy(&check_for_self_loops);
+            igraph_vector_destroy(&cluster_member);
+            IGRAPH_FINALLY_CLEAN(4);
+
+            return IGRAPH_SUCCESS;
+        }
+    }
+
+    /* assuming that cluster count is not greater than 1 now */
+    /* however, we need to check for self loops still*/
+
+    odd = 0;
+
+    IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+
+    IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+    vsize = igraph_vector_size(&degree);
+    for (i = 0; i < vsize; i++) {
+        if (((long int) VECTOR(degree)[i]) % 2 == 1) odd++;
+    }
+
+    if (odd > 2) {
+        *has_path = 0;
+        *has_cycle = 0;
+    } else if (odd == 2) {
+        *has_path = 1;
+        *has_cycle = 0;
+    } else {
+        *has_path = 1;
+        *has_cycle = 1;
+    }
+
+    vsize = igraph_vector_size(&degree);
+    for (i = 0; i < vsize; i++) {
+        if ((*has_cycle && ((long int) VECTOR(degree)[i]) > 0) || (!*has_cycle && ((long int) VECTOR(degree)[i]) %2 == 1)) {
+            *start_of_path = i;
+            break;
+        }
+    }
+
+    igraph_vector_destroy(&csize);
+    igraph_vector_destroy(&check_for_self_loops);
+    igraph_vector_destroy(&degree);
+    igraph_inclist_destroy(&il);
+    igraph_vector_destroy(&cluster_member);
+    IGRAPH_FINALLY_CLEAN(5);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+static int igraph_i_is_eulerian_directed(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle, igraph_integer_t *start_of_path) {
+    igraph_bool_t res_weak;
+    igraph_integer_t incoming_excess, outgoing_excess, n;
+    long int i, j, vsize;
+    long int cluster_count;
+    igraph_vector_t out_degree, in_degree, csize_weak, check_for_self_loops, cluster_member;
+    igraph_inclist_t il;
+
+    if (!graph) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    }
+
+    n = igraph_vcount(graph);
+
+    if (igraph_ecount(graph) == 0 || n <= 1) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    }
```

Comment:
Whenever the function returns with `has_path` being true, it should not leave `start_of_path` un-set.

Reply:
Just fixed this.